File: contest/ICPC/2019/北美洲/美国中部/data/sumandproduct/submissions/accepted/bowen.py
#!/usr/bin/env python2

import sys

MAX_SUM = 2e14

n = int(sys.stdin.readline())
arr = [int(x) for x in sys.stdin.readline().split()]
next_one = [-1] * (n + 1)
for i in range(n-1, -1, -1):
  if arr[i] == 1:
    next_one[i] = max(next_one[i + 1], i)

ans = 0
for i in range(n):
  sum, prod, j = arr[i], arr[i], i+1
  while j < n:
    if arr[j] == 1:
      cnt_one = next_one[j] - j + 1
      if 1 <= prod - sum <= cnt_one: ans += 1
      sum += cnt_one
      j = next_one[j]
    else:
      sum += arr[j]
      prod *= arr[j]
      if prod > MAX_SUM: break
      if sum == prod: ans += 1
    j += 1
        
print (ans)

File: contest/ICPC/2019/北美洲/美国中部/data/sumandproduct/submissions/accepted/bowen.cpp
#include <iostream>
#include <cstdio>
#include <cmath>
#include <cstring>
#include <algorithm>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <set>
#include <map>
#include <sstream>
#include <complex>
#include <ctime>
#include <cassert>
#include <functional>

using namespace std;

typedef long long ll;
typedef vector<int> VI;
typedef pair<int, int> PII;

#define REP(i,s,t) for(int i=(s);i<(t);i++)
#define FILL(x,v) memset(x,v,sizeof(x))

const ll INF = (ll)2E14;
#define MAXN 200005

int endOne[MAXN], a[MAXN];
int main() {
	int n;
	cin >> n;
	REP(i, 0, n) cin >> a[i];
	for (int i = n - 1; i >= 0; i--) {
		if (a[i] == 1) endOne[i] = max(i, endOne[i + 1]);
		else endOne[i] = -1;
	}
	ll ans = 0;
	REP(i, 0, n) {
		ll sum = a[i], prod = a[i];
		REP(j, i+1, n) {
			if (a[j] == 1) {
				int cntOne = endOne[j] - j + 1;
				if (prod > sum && cntOne >= prod - sum) ans++;
				sum += cntOne;
				j = endOne[j];
			} else {
				sum += a[j];
				if (prod > INF / a[j]) break;
				prod *= a[j];

				if (sum == prod) ans++;
			}
		}
	}
	cout << ans << endl;
	return 0;
}